compiler/src/iree/compiler/GlobalOptimization/DecomposeLowParallelismAttentionPass.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/Dialect/LinalgExt/Utils/IndexingUtils.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+#define GEN_PASS_DEF_DECOMPOSELOWPARALLELISMATTENTIONPASS
+#include "iree/compiler/GlobalOptimization/Passes.h.inc"
+
+using namespace IREE::LinalgExt;
+
+namespace {
+
+const int64_t kDynamicDimMaxValue = 65536; // 2^16
+/// TODO: The M heuristic is a conservative choice I(Groverkss) made.
+/// We ideally need more data to determine when it is good to
+/// decompose.
+const int64_t kMParallelismThreshold = 4;
+/// The K1 heuristic is simply based on the fact that if we have a tile size
+/// bigger than 256, we should probably be tiling that dimension. Non decomposed
+/// attention is not tilable along K1 diomension.
+const int64_t kK1ParallelismThreshold = 256;
+
+static int64_t calculateSize(ArrayRef<int64_t> bounds, ArrayRef<int64_t> dims) {
+  int64_t size = 1;
+  for (int64_t dim : dims) {
+    if (ShapedType::isDynamic(bounds[dim])) {
+      // TODO: We should query range information to find the maximum value
+      // possible instead.
+      size *= kDynamicDimMaxValue;
```

Comment:
does it makes sense to just do `return ShapedType::kDynamic` if we find a dynamic shape?

Reply:
actually not sure about shapedType::kDYnamic since it may be -1 or something like thatm but maybe we can do an early `return kDynamicDimMaxValue`